### ISSUE TITLE
esp8266: RTC wraps around in 8 hours

### ISSUE THREAD

**Issue by puuu** [micropython/micropython#2097]
The rtc of the esp8266 loops in 8 hour periods. I can reproduce this with `time.localtime()`, `time.time()`, `machine.RTC().datetime()` on a WROOM-02 and HUZZAH with a compiled micropython version from 2016-05-17. The behaviour seems to be independent from the date.

Here is a simple script to reproduce this behaviour:

```python
import ntptime, time, machine

ntptime.settime()

last = -1
while True:
    epoche = time.time()
    tim = time.localtime()
    if last != tim[3]:
        last = tim[3]
        print("time.localtime()", tim)
        print("time.time()", epoche)
        print("time.localtime(time.time())", time.localtime(epoche))
        print("machine.RTC().datetime()", machine.RTC().datetime())
    time.sleep_ms(100)
```

The output is:
```
(2016, 5, 19, 9, 54, 28, 3, 140)
time.localtime() (2016, 5, 19, 9, 54, 28, 3, 140)
time.time() 516966868
time.localtime(time.time()) (2016, 5, 19, 9, 54, 28, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 9, 54, 28, 4)
time.localtime() (2016, 5, 19, 10, 0, 0, 3, 140)
time.time() 516967200
time.localtime(time.time()) (2016, 5, 19, 10, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 10, 0, 0, 8)
time.localtime() (2016, 5, 19, 11, 0, 0, 3, 140)
time.time() 516970800
time.localtime(time.time()) (2016, 5, 19, 11, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 11, 0, 0, 97)
time.localtime() (2016, 5, 19, 12, 0, 0, 3, 140)
time.time() 516974400
time.localtime(time.time()) (2016, 5, 19, 12, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 12, 0, 0, 5)
time.localtime() (2016, 5, 19, 13, 0, 0, 3, 140)
time.time() 516978000
time.localtime(time.time()) (2016, 5, 19, 13, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 13, 0, 0, 83)
time.localtime() (2016, 5, 19, 14, 0, 0, 3, 140)
time.time() 516981600
time.localtime(time.time()) (2016, 5, 19, 14, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 14, 0, 0, 55)
time.localtime() (2016, 5, 19, 15, 0, 0, 3, 140)
time.time() 516985200
time.localtime(time.time()) (2016, 5, 19, 15, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 15, 0, 0, 50)
time.localtime() (2016, 5, 19, 16, 0, 0, 3, 140)
time.time() 516988800
time.localtime(time.time()) (2016, 5, 19, 16, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 16, 0, 0, 61)
time.localtime() (2016, 5, 19, 8, 37, 9, 3, 140)
time.time() 516962229
time.localtime(time.time()) (2016, 5, 19, 8, 37, 9, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 8, 37, 9, 248)
time.localtime() (2016, 5, 19, 9, 0, 0, 3, 140)
time.time() 516963600
time.localtime(time.time()) (2016, 5, 19, 9, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 9, 0, 0, 44)
time.localtime() (2016, 5, 19, 10, 0, 0, 3, 140)
time.time() 516967200
time.localtime(time.time()) (2016, 5, 19, 10, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 10, 0, 0, 17)
time.localtime() (2016, 5, 19, 11, 0, 0, 3, 140)
time.time() 516970800
time.localtime(time.time()) (2016, 5, 19, 11, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 11, 0, 0, 81)
time.localtime() (2016, 5, 19, 12, 0, 0, 3, 140)
time.time() 516974400
time.localtime(time.time()) (2016, 5, 19, 12, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 12, 0, 0, 97)
time.localtime() (2016, 5, 19, 13, 0, 0, 3, 140)
time.time() 516978000
time.localtime(time.time()) (2016, 5, 19, 13, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 13, 0, 0, 18)
time.localtime() (2016, 5, 19, 14, 0, 0, 3, 140)
time.time() 516981600
time.localtime(time.time()) (2016, 5, 19, 14, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 14, 0, 0, 17)
time.localtime() (2016, 5, 19, 15, 0, 0, 3, 140)
time.time() 516985200
time.localtime(time.time()) (2016, 5, 19, 15, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 15, 0, 0, 41)
time.localtime() (2016, 5, 19, 16, 0, 0, 3, 140)
time.time() 516988800
time.localtime(time.time()) (2016, 5, 19, 16, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 16, 0, 0, 94)
time.localtime() (2016, 5, 19, 8, 37, 9, 3, 140)
time.time() 516962229
time.localtime(time.time()) (2016, 5, 19, 8, 37, 9, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 8, 37, 9, 310)
time.localtime() (2016, 5, 19, 9, 0, 0, 3, 140)
time.time() 516963600
time.localtime(time.time()) (2016, 5, 19, 9, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 9, 0, 0, 36)
time.localtime() (2016, 5, 19, 10, 0, 0, 3, 140)
time.time() 516967200
time.localtime(time.time()) (2016, 5, 19, 10, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 10, 0, 0, 68)
time.localtime() (2016, 5, 19, 11, 0, 0, 3, 140)
time.time() 516970800
time.localtime(time.time()) (2016, 5, 19, 11, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 11, 0, 0, 52)
time.localtime() (2016, 5, 19, 12, 0, 0, 3, 140)
time.time() 516974400
time.localtime(time.time()) (2016, 5, 19, 12, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 12, 0, 0, 89)
time.localtime() (2016, 5, 19, 13, 0, 0, 3, 140)
time.time() 516978000
time.localtime(time.time()) (2016, 5, 19, 13, 0, 0, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 13, 0, 0, 51)
```

You see, after 16:00 always comes 8:37 and the date will not change. Stopping the script and running `ntptime.settime()` will print `(2016, 5, 20, 4, 28, 34, 4, 141)`, which is correct.

Another run is:
```
(2016, 5, 20, 4, 28, 34, 4, 141)
time.localtime() (2016, 5, 20, 4, 28, 34, 4, 141)
time.time() 517033714
time.localtime(time.time()) (2016, 5, 20, 4, 28, 34, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 4, 28, 34, 4)
time.localtime() (2016, 5, 20, 5, 0, 0, 4, 141)
time.time() 517035600
time.localtime(time.time()) (2016, 5, 20, 5, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 5, 0, 0, 23)
time.localtime() (2016, 5, 20, 6, 0, 0, 4, 141)
time.time() 517039200
time.localtime(time.time()) (2016, 5, 20, 6, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 6, 0, 0, 16)
time.localtime() (2016, 5, 20, 7, 0, 0, 4, 141)
time.time() 517042800
time.localtime(time.time()) (2016, 5, 20, 7, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 7, 0, 0, 58)
time.localtime() (2016, 5, 19, 23, 42, 13, 3, 140)
time.time() 517016533
time.localtime(time.time()) (2016, 5, 19, 23, 42, 13, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 23, 42, 13, 806)
time.localtime() (2016, 5, 20, 0, 0, 0, 4, 141)
time.time() 517017600
time.localtime(time.time()) (2016, 5, 20, 0, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 0, 0, 0, 76)
time.localtime() (2016, 5, 20, 1, 0, 0, 4, 141)
time.time() 517021200
time.localtime(time.time()) (2016, 5, 20, 1, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 1, 0, 0, 45)
time.localtime() (2016, 5, 20, 2, 0, 0, 4, 141)
time.time() 517024800
time.localtime(time.time()) (2016, 5, 20, 2, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 2, 0, 0, 59)
time.localtime() (2016, 5, 20, 3, 0, 0, 4, 141)
time.time() 517028400
time.localtime(time.time()) (2016, 5, 20, 3, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 3, 0, 0, 35)
time.localtime() (2016, 5, 20, 4, 0, 0, 4, 141)
time.time() 517032000
time.localtime(time.time()) (2016, 5, 20, 4, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 4, 0, 0, 45)
time.localtime() (2016, 5, 20, 5, 0, 0, 4, 141)
time.time() 517035600
time.localtime(time.time()) (2016, 5, 20, 5, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 5, 0, 0, 15)
time.localtime() (2016, 5, 20, 6, 0, 0, 4, 141)
time.time() 517039200
time.localtime(time.time()) (2016, 5, 20, 6, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 6, 0, 0, 14)
time.localtime() (2016, 5, 20, 7, 0, 0, 4, 141)
time.time() 517042800
time.localtime(time.time()) (2016, 5, 20, 7, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 7, 0, 0, 62)
time.localtime() (2016, 5, 19, 23, 42, 13, 3, 140)
time.time() 517016533
time.localtime(time.time()) (2016, 5, 19, 23, 42, 13, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 23, 42, 13, 788)
time.localtime() (2016, 5, 20, 0, 0, 0, 4, 141)
time.time() 517017600
time.localtime(time.time()) (2016, 5, 20, 0, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 0, 0, 0, 15)
time.localtime() (2016, 5, 20, 1, 0, 0, 4, 141)
time.time() 517021200
time.localtime(time.time()) (2016, 5, 20, 1, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 1, 0, 0, 35)
time.localtime() (2016, 5, 20, 2, 0, 0, 4, 141)
time.time() 517024800
time.localtime(time.time()) (2016, 5, 20, 2, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 2, 0, 0, 19)
time.localtime() (2016, 5, 20, 3, 0, 0, 4, 141)
time.time() 517028400
time.localtime(time.time()) (2016, 5, 20, 3, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 3, 0, 0, 99)
time.localtime() (2016, 5, 20, 4, 0, 0, 4, 141)
time.time() 517032000
time.localtime(time.time()) (2016, 5, 20, 4, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 4, 0, 0, 39)
time.localtime() (2016, 5, 20, 5, 0, 0, 4, 141)
time.time() 517035600
time.localtime(time.time()) (2016, 5, 20, 5, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 5, 0, 0, 75)
time.localtime() (2016, 5, 20, 6, 0, 0, 4, 141)
time.time() 517039200
time.localtime(time.time()) (2016, 5, 20, 6, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 6, 0, 0, 3)
time.localtime() (2016, 5, 20, 7, 0, 0, 4, 141)
time.time() 517042800
time.localtime(time.time()) (2016, 5, 20, 7, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 7, 0, 0, 31)
time.localtime() (2016, 5, 19, 23, 42, 13, 3, 140)
time.time() 517016533
time.localtime(time.time()) (2016, 5, 19, 23, 42, 13, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 23, 42, 13, 815)
time.localtime() (2016, 5, 20, 0, 0, 0, 4, 141)
time.time() 517017600
time.localtime(time.time()) (2016, 5, 20, 0, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 0, 0, 0, 95)
time.localtime() (2016, 5, 20, 1, 0, 0, 4, 141)
time.time() 517021200
time.localtime(time.time()) (2016, 5, 20, 1, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 1, 0, 0, 61)
time.localtime() (2016, 5, 20, 2, 0, 0, 4, 141)
time.time() 517024800
time.localtime(time.time()) (2016, 5, 20, 2, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 2, 0, 0, 3)
time.localtime() (2016, 5, 20, 3, 0, 0, 4, 141)
time.time() 517028400
time.localtime(time.time()) (2016, 5, 20, 3, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 3, 0, 0, 62)
time.localtime() (2016, 5, 20, 4, 0, 0, 4, 141)
time.time() 517032000
time.localtime(time.time()) (2016, 5, 20, 4, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 4, 0, 0, 69)
time.localtime() (2016, 5, 20, 5, 0, 0, 4, 141)
time.time() 517035600
time.localtime(time.time()) (2016, 5, 20, 5, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 5, 0, 0, 35)
time.localtime() (2016, 5, 20, 6, 0, 0, 4, 141)
time.time() 517039200
time.localtime(time.time()) (2016, 5, 20, 6, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 6, 0, 0, 36)
time.localtime() (2016, 5, 20, 7, 0, 0, 4, 141)
time.time() 517042800
time.localtime(time.time()) (2016, 5, 20, 7, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 7, 0, 0, 97)
time.localtime() (2016, 5, 19, 23, 42, 13, 3, 140)
time.time() 517016533
time.localtime(time.time()) (2016, 5, 19, 23, 42, 13, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 23, 42, 13, 775)
time.localtime() (2016, 5, 20, 0, 0, 0, 4, 141)
time.time() 517017600
time.localtime(time.time()) (2016, 5, 20, 0, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 0, 0, 0, 55)
time.localtime() (2016, 5, 20, 1, 0, 0, 4, 141)
time.time() 517021200
time.localtime(time.time()) (2016, 5, 20, 1, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 1, 0, 0, 55)
time.localtime() (2016, 5, 20, 2, 0, 0, 4, 141)
time.time() 517024800
time.localtime(time.time()) (2016, 5, 20, 2, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 2, 0, 0, 75)
time.localtime() (2016, 5, 20, 3, 0, 0, 4, 141)
time.time() 517028400
time.localtime(time.time()) (2016, 5, 20, 3, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 3, 0, 0, 30)
time.localtime() (2016, 5, 20, 4, 0, 0, 4, 141)
time.time() 517032000
time.localtime(time.time()) (2016, 5, 20, 4, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 4, 0, 0, 89)
time.localtime() (2016, 5, 20, 5, 0, 0, 4, 141)
time.time() 517035600
time.localtime(time.time()) (2016, 5, 20, 5, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 5, 0, 0, 5)
time.localtime() (2016, 5, 20, 6, 0, 0, 4, 141)
time.time() 517039200
time.localtime(time.time()) (2016, 5, 20, 6, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 6, 0, 0, 86)
time.localtime() (2016, 5, 20, 7, 0, 0, 4, 141)
time.time() 517042800
time.localtime(time.time()) (2016, 5, 20, 7, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 7, 0, 0, 70)
time.localtime() (2016, 5, 19, 23, 42, 13, 3, 140)
time.time() 517016533
time.localtime(time.time()) (2016, 5, 19, 23, 42, 13, 3, 140)
machine.RTC().datetime() (2016, 5, 19, 3, 23, 42, 13, 783)
time.localtime() (2016, 5, 20, 0, 0, 0, 4, 141)
time.time() 517017600
time.localtime(time.time()) (2016, 5, 20, 0, 0, 0, 4, 141)
machine.RTC().datetime() (2016, 5, 20, 4, 0, 0, 0, 25)
```

Here, after 7:00 always comes 23:42 and the day will jump one day backward. Again, stopping the script and running `ntptime.settime()` will print `(2016, 5, 21, 14, 31, 39, 5, 142)`, which is correct.

**Comment by puuu** [comment]
This seems to be a bug in the esp SDK, as discussed here: http://bbs.espressif.com/viewtopic.php?t=107
`system_get_rtc_time()` is only uint32 which results in an overflow after about 7.45h.

I think, we should detect the overflow, by storing the rtc time in the rtc memory. At the moment, I am testing a fix. If it runs 24 hours without problems, I will open a pull request for further discussion.

**Comment by puuu** [comment]
Fixed with #2107.
